internal/repositories/github/service.go

```diff
@@ -70,6 +92,63 @@ func NewRepositoryService(
 	}
 }
 
+func (r *repositoryService) CreateRepository(
+	ctx context.Context,
+	client ghclient.GitHubRepoClient,
+	projectID uuid.UUID,
+	repo *pb.UpstreamRepositoryRef,
+) (*pb.Repository, error) {
+	// get information about the repo from GitHub, and ensure it exists
+	githubRepo, err := client.GetRepository(ctx, repo.Owner, repo.Name)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving repo from github: %w", err)
+	}
+
+	// skip if this is a private repo, and private repos are not enabled
+	if githubRepo.GetPrivate() && !features.ProjectAllowsPrivateRepos(ctx, r.store, projectID) {
+		return nil, ErrPrivateRepoForbidden
+	}
+
+	// create a webhook to capture events from the repository
+	hookUUID, githubHook, err := r.webhookManager.CreateWebhook(ctx, client, repo.Owner, repo.Name)
+	if err != nil {
+		return nil, fmt.Errorf("error creating webhook in repo: %w", err)
+	}
+
+	// insert the repository into the DB
+	dbID, pbRepo, err := r.persistRepository(
+		ctx,
+		githubRepo,
+		githubHook,
+		hookUUID,
+		projectID,
+	)
+	if err != nil {
+		log.Printf("error creating repository '%s/%s' in database: %v", repo.Owner, repo.Name, err)
+		// Attempt to clean up the webhook we created earlier. This is a
+		// best-effort attempt: If it fails, the customer either has to delete
+		// the hook manually, or it will be deleted the next time the customer
+		// attempts to register a repo.
+		cleanupErr := r.webhookManager.DeleteWebhook(ctx, client, repo.Owner, repo.Name, *githubHook.ID)
+		if cleanupErr != nil {
+			log.Printf("error deleting new webhook: %v", cleanupErr)
+		}
+		return nil, fmt.Errorf("error creating repository in database: %w", err)
+	}
+
+	// publish a reconciling event for the registered repositories
+	if err = r.pushReconcilerEvent(pbRepo, projectID); err != nil {
+		return nil, err
+	}
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = ghprovider.Github
+	logger.BusinessRecord(ctx).Project = projectID
+	logger.BusinessRecord(ctx).Repository = dbID
```

Comment:
Can we move these up earlier in the function to associate more errors with projects and repos?